xarray/tests/duckarrays/base/strategies.py

```diff
@@ -0,0 +1,104 @@
+import hypothesis.extra.numpy as npst
+import hypothesis.strategies as st
+
+import xarray as xr
+
+
+def shapes(ndim):
+    return npst.array_shapes(min_dims=ndim, max_dims=ndim)
+
+
+dtypes = (
+    npst.integer_dtypes()
+    | npst.unsigned_integer_dtypes()
+    | npst.floating_dtypes()
+    | npst.complex_number_dtypes()
+)
+
+
+def numpy_array(shape=None):
+    if shape is None:
+        shape = shapes()
+    return npst.arrays(dtype=dtypes, shape=shape)
+
+
+def create_dimension_names(ndim):
+    return [f"dim_{n}" for n in range(ndim)]
+
+
+@st.composite
+def variable(draw, create_data, *, sizes=None):
+    if sizes is not None:
+        dims, shape = zip(*draw(sizes).items())
+    else:
+        dims = draw(st.lists(st.text(min_size=1), max_size=4, unique=True).map(tuple))
+        ndim = len(dims)
+        shape = draw(shapes(ndim))
+
+    data = draw(create_data(shape))
+    return xr.Variable(dims, data)
+
+
+@st.composite
+def data_array(draw, create_data):
+    name = draw(st.none() | st.text(min_size=1))
+
+    dims = draw(st.lists(elements=st.text(min_size=1), max_size=4, unique=True))
+    shape = draw(shapes(len(dims)))
```

Comment:
A more idiomatic pattern would be to draw a list of `(dim_name, dim_siz)` tuples - that way Hypothesis internals will be aware of the correspondence and can take advantage when generating or shrinking.

```python
st.lists(
    elements=st.tuples(st.text(min_size=1), st.integers(1, 10)),
    max_size=4,
    unique_by=lambda x: x[0]
)
```

In itself I'd only expect a small perf gain, but when working with arrays these things can compound.